src/lib/data.nix

```diff
@@ -56,8 +65,11 @@ let
         preferLocalBuild = true;
         allowSubstitutes = false;
       } ''
-        python ${pkgs.styx}/share/styx/tools/parser.py < ${fileData.path} > $out
-        '';
+  ${if preprocess ? "${markupType}" then
+        "${preprocess."${markupType}"} ${fileData.path} > preprocessed"
+  else "cp ${fileData.path} preprocessed"}
+        python ${pkgs.styx}/share/styx/tools/parser.py < preprocessed > $out
+  '';
```

Comment:
This should not be needed, the running order is parsing -> processing.
Parsing take a markup file with styx specific markup and turn it in a nix attribute set, containing `pages`, `content` and so on, that are totally cleaned of styx specific markup;
These attributes are then fed to `markupToHtml` to convert it to html in https://github.com/styx-static/styx/blob/master/src/lib/data.nix#L62-L67.

So just setting a `command.orgmode` should be enough.

Reply:
"Parsing take a markup file with styx specific markup and turn it in a nix attribute set, containing pages, content and so on"

The preprocessing is creating the styx specific markup out of the org standard tags. eg. in org you can use #+title: title to set the document title. This is compatible with the html exporter, with the latex exproter, beamer, and any other exporter, so by writing an org file you just assume it's the same. The preprocess takes care of creating the styx "title", "tags" and "draft" attributes.

Without the first step, one should manually write those elements in a non-org sintax (the styx one)

Reply:
Ok, I understand better now the need for preprocessing.

> The preprocess takes care of creating the styx "title", "tags" and "draft" attributes.

Metadata is a totally free format, there are case where other attributes are used, eg author, date or any arbitratry attribute like [level or banner](https://github.com/styx-static/styx/blob/master/src/scaffold/sample-data/posts/2016-09-16-themes.md). Will the preprocessing approach allow to set any attribute?

Reply:
Also is the preprocessing step compatible with [embedded nix](https://styx-static.github.io/styx-site/documentation/index.html#data.embedded-nix)? Having embedded nix is nice as it allows to do  in site links and template calls from the content. (I guess this would bring the problem of non compatible orgmode content again)

Reply:
I just had a thought, but in case it is hard to find a good tradeoff, what about using a flexible approach:

- If there is no styx metadata attribute at the top of the file, generate it like done in the preprocessing
- If there is a styx metadata attribute at the top of the file, just use it and don't generate one

This could fit both use cases, what do you think?

Reply:
For the compatibility with embedded nix: it works! (and I love the fact that it is. I'm not doing anything, but luckily  this does not conflict with the org sintax. in org you have {{{macros}}} with 3 )
example:
```org
\{{ toString (2 + 2) }} returns: {{ toString (2 + 2) }}
```
becomes:
```nix
{{ toString (2 + 2) }} returns: 4
```

Updated example, here again: https://blog.nixo.xyz/posts/2018-01-26-org-mode.html

Reply:
About the metadata: if _any_ metadata available, the lowest friction I can imagine is:
export the default fields, that are title, date, author, email, language, creator, by default, and then have a custom keyword like:
```org
#+STYX: option:value
``` where one can set styx-specific ones. It seems to me the lowest friction/best default behavior. If a different behavior is needed, it's simply possible to replace the org-preprocess/org-compile functions in the `.el` file
```org
#+TITLE: 2018-01-26-org-mode
#+DATE: <2018-02-01 Thu>
#+AUTHOR: Nicolò Balzarotti
#+EMAIL: anothersms@gmail.com
#+LANGUAGE: en
#+CREATOR: Emacs 25.3.1 (Org mode 9.1.6)
```

Reply:
Sorry I read the last comment  ("This could fit both use cases, what do you think?") only now. I'm not sure if this approach is better than the previous. letting #+STYX: option:value override params like title, author and so on is the best approach I can think of, so that the file continue to be export-able in any other org-supported format without compromises (ie, having a {{-- at the top)

Reply:
The `#+STYX: option:value` way is great!